include/sys/onoff.h

```diff
@@ -0,0 +1,606 @@
+					    struct k_poll_signal *sigp)
+{
+#ifdef CONFIG_POLL
+	__ASSERT_NO_MSG(cli != NULL);
+	__ASSERT_NO_MSG(sigp != NULL);
+
+	*cli = (struct onoff_client){
+		.async = {
+			.signal = sigp,
+		},
+		.flags = ONOFF_CLIENT_NOTIFY_SIGNAL,
+	};
+#endif /* CONFIG_POLL */
+}
+
+/**
+ * @brief Initialize an on-off client to be used for a callback
+ * operation notification.
+ *
+ * Clients that use this initialization will be notified of the
+ * completion of operations submitted through on-off service API
+ * through the provided callback.  Note that callbacks may be invoked
+ * from various contexts depending on the specific service; see
+ * @ref onoff_client_callback.
+ *
+ * @param cli pointer to the client state object.
+ *
+ * @param handler a function pointer to use for notification.
+ *
+ * @param user_data an opaque pointer passed to the handler to provide
+ * additional context.
+ */
+static inline void onoff_client_init_callback(struct onoff_client *cli,
+					      onoff_client_callback handler,
+					      void *user_data)
+{
+	__ASSERT_NO_MSG(cli != NULL);
+	__ASSERT_NO_MSG(handler != NULL);
+
+	*cli = (struct onoff_client){
+		.async = {
+			.callback = {
+				.handler = handler,
+				.user_data = user_data,
+			},
+		},
+		.flags = ONOFF_CLIENT_NOTIFY_CALLBACK,
+	};
+}
+
+/**
+ * @brief Request a reservation to use an on-off service.
+ *
+ * The return value indicates the success or failure of an attempt to
+ * initiate an operation to request the resource be made available.
+ * If initiation of the operation succeeds the result of the request
+ * operation is provided through the configured client notification
+ * method, possibly before this call returns.
+ *
+ * Note that the call to this function may succeed in a case where the
+ * actual request fails.  Always check the operation completion
+ * result.
+ *
+ * As a specific example: A call to this function may succeed at a
+ * point while the service is still transitioning to off due to a
+ * previous call to onoff_release().  When the transition completes
+ * the service would normally start a transition to on.  However, if
+ * the transition to off completed in a non-thread context, and the
+ * transition to on can sleep, the transition cannot be started and
+ * the request will fail with `-EWOULDBLOCK`.
+ *
+ * @param srv the service that will be used.
+ *
+ * @param cli a non-null pointer to client state providing
+ * instructions on synchronous expectations and how to notify the
+ * client when the request completes.  Behavior is undefined if client
+ * passes a pointer object associated with an incomplete service
+ * operation.
+ *
+ * @retval Non-negative on successful (initiation of) request
+ * @retval -EIO if service has recorded an an error
+ * @retval -EINVAL if the parameters are invalid
+ * @retval -EAGAIN if the reference count would overflow
+ * @retval -EWOULDBLOCK if the function was invoked from non-thread
```

Comment:
there is another reason for that: if request was called during to_off transition and stop notification was called from interrupt context and start sleeps. I know that it will be quite complex for the user so ideally there should be provided a pattern how to handle it (in doc or dox). 

is that there right way of handling -EWOULDBLOCK?
```
err = onoff_request(srv,cli);
while (err == -EWOULDBLOCK && !(k_is_in_isr() || k_is_pre_kernel())) {
  /* if we are in thread context and service cannot be started it
   * means that we hit to off transition. Try again. */
    err = onoff_request(srv_cli);
}
if (err < 0) {
  return err;
}
```

Reply:
> there is another reason for that [`-EWOULDBLOCK`]: if request was called during to_off transition and stop notification was called from interrupt context and start sleeps.

No, if `onoff_request()` is called while the service is being turned off then this call will return success.  The asynchronous operation result would be `-EWOULDBLOCK` if the off transition completes in non-thread context.

Reply:
right, i confused return value with operation result which may be `-EWOULDBLOCK` in case when `onoff_stop_notify` is called from irq context, start sleeps and request came while stopping.

Still, it seems that fail safe handling is quite complex and it would be good to show patterns.

Here is something i came up with assuming that service has `ONOFF_SERVICE_START_SLEEPS` and `ONOFF_CLIENT_NOTIFY_SIGNAL` client.
```
bool reset_done;
do {
  err = onoff_request(srv, cli);
  if (err < 0) {
    break;
  }
  k_poll(..);
  k_poll_signal_check(cli->async.signal, &signaled, &result);

  if (result == 0) {
   break;
  } else if (result != -EWOULDBLOCK && !reset_done) {
     /* request failed, if reset is supported, attempt to reset? */
     err = onoff_service_reset(srv, cli);
      if (err < 0) {
        break;
      }
      k_poll(..);
      k_poll_signal_check(cli->async.signal, &signaled, &result);
      if (result < 0) {
        break;
      }
      reset_done = true; /* prevent forever loop if service is corrupted. */
  } else {
     /* WOULDBLOCK - corner case, start failed because attempted to start from wrong context*/
  }
} while (true); 
```

Reply:
OK, depending on what that's meant to do it's not right in several ways.  I'll work up a sample to show how various situations need to be handled.

Reply:
>  it's not right in several ways

that only proves that sample is needed 😄

Reply:
Eh, maybe.  Some of the problems would be resolved by testing the code, but more generally the service reset API is present only as support for edge cases, where it is possible to recover from a service error.  I have yet to identify a use case where such recovery is possible, so I'd rather wait until we have something to work with rather than put a guess as to what it could look like in documentation and a sample.

https://gist.github.com/pabigot/6176d58fc60b6f3f870209368a6caae4 shows something that I think would work and documents the edge cases, though I haven't even compiled it.  If you want more than this I won't be able to get to it until the end of next week.